### PR TITLE
Mm44851 get involved

### PR DESCRIPTION
**Pull request by trilopin**
#### Summary
Get involved flow.

If a viewer user wants to get involved in a run, can click the get involved button and:
- A: if the channel is accessible will join it directly
- B: a message will be sent to the channel otherwise

Copies are not final.

**Demo B**
https://user-images.githubusercontent.com/4096774/177756194-1e407760-1286-48c9-9d91-a59fe78aebd6.mp4

**Demo A**
https://user-images.githubusercontent.com/4096774/177756213-9fe18668-b2f1-452d-8b19-a6d6cb36dd47.mp4

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-44851

#### Checklist
- [ ] ~~Telemetry updated~~
- [X] Gated by experimental feature flag
- [x] Unit tests updated

### CODE REVIEW

**Review by agarciamontoro — COMMENTED** [review]
I think the approach makes total sense. I may add a comment to the `channel === null` explaining we assume that that means the channel is private. But apart from that, I think everything looks good!

**Comment by trilopin** [comment]
cc @justinegeffen , here we have the efforts for the get involved feature. Feel free to suggest awesome copies right here.

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
This looks great! I left a bunch of comments, but mostly nits and typos. I have a general UX comment, though: Should we also show a confirmation modal when the user will directly join the channel? Something like "Do you want to be added to this run's channel, ~channelname?" People may not be aware that "Get involved" means joining a channel.

Also, future improvements for v2: Wen someone requests to get involved to a private channel and a message is sent to the channel, we may render a custom post with an `Invite` button that automatically invites the member to the channel.

File: server/app/playbook_run_service.go [hunk 1]
```diff
@@ -2696,6 +2696,29 @@ func (s *PlaybookRunServiceImpl) RequestUpdate(playbookRunID, requesterID string
...
+func (s *PlaybookRunServiceImpl) RequestGetInvolved(playbookRunID, requesterID string) error {
+	playbookRun, err := s.store.GetPlaybookRun(playbookRunID)
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve playbook run")
+	}
+
+	requesterUser, err := s.pluginAPI.User.Get(requesterID)
+	if err != nil {
+		return errors.Wrap(err, "failed to get requester user")
+	}
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/app/playbook_run_service.go:2709]: do you mean to check in addition to the frontend check? because the webapp button won't be shown if the user is already a participant

[the remark this answers is not in the document]
Comment (agarciamontoro) [reply server/app/playbook_run_service.go:2709]: Yes, I think we should check it here, too, because a user could still do it through the API. It's probably an edge case, but I think it's an easy enough check, so I'd say it's worth doing it.

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 2]
```diff
@@ -401,9 +425,98 @@ describe('runs > run details page > header', () => {
...
+                    // * Assert modal is not showed
+                    cy.get('#confirmModal').should('not.exist');
+
+                    // # Login as testUser
+                    cy.apiLogin(testUser).then(() => {
+                        // # Visit the channel run
+                        cy.visit(`${testTeam.name}/channels/${playbookRunChannelName}`);
+
+                        // * Assert message has not been sent
+                        cy.get('#postListContent').should('not.contain', 'wants to get involved in this run.');
```
[the remark this answers is not in the document]
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:459]: didn't know, it's a useful helper

**Comment by trilopin** [comment]
> This looks great! I left a bunch of comments, but mostly nits and typos. I have a general UX comment, though: Should we also show a confirmation modal when the user will directly join the channel? Something like "Do you want to be added to this run's channel, ~channelname?" People may not be aware that "Get involved" means joining a channel.

I've mixed feelings about that. Modals are educative but also stop the flow. Since the user could leave the channel I would start like this unless we hear about tons of accidental channel join. Additionally, the modal to confirm the request also has a motivation to avoid flooding the channel (also could be done through a debounce tho). 

 
> Also, future improvements for v2: Wen someone requests to get involved to a private channel and a message is sent to the channel, we may render a custom post with an `Invite` button that automatically invites the member to the channel.

Couldn't agree more. Nice topic to revisit after checking usage metrics.

**Review by agarciamontoro — APPROVED** [review]
Thanks for addressing all my comments! :)

> I would start like this unless we hear about tons of accidental channel join

Ok, let's go with that!

File: server/app/playbook_run_service.go [hunk 3]
```diff
@@ -2696,6 +2696,35 @@ func (s *PlaybookRunServiceImpl) RequestUpdate(playbookRunID, requesterID string
...
+	member, err := s.pluginAPI.Channel.GetMember(playbookRun.ChannelID, requesterID)
+	if err == nil && member != nil {
+		return errors.New("user is already involved")
+	}
+
+	T := i18n.GetUserTranslations(requesterUser.Locale)
+	data := map[string]interface{}{
+		"Name": requesterUser.Username,
+	}
+	if _, err = s.poster.PostMessage(playbookRun.ChannelID, T("app.user.run.request_get_involved", data)); err != nil {
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/app/playbook_run_service.go:2721]: is your concern related to this server code or the overall behavior?

I mean, by using the feature as a user, message and sender are fixed (copy and playbooks bot) and could not be arbitrary. So I guess it's more about this code at server side.

What kind of bad use are you worried about?

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 4]
```diff
@@ -401,9 +425,98 @@ describe('runs > run details page > header', () => {
...
+                    // * Assert thst button is showed
+                    getHeader().findByText('Get involved').should('be.visible');
+                });
+
+                it('click button to show modal and cancel', () => {
+                    // * Assert that component is rendered
+                    getHeader().findByText('Get involved').should('be.visible');
+
+                    // # Wait for useChannel
+                    cy.wait(500);
```
[the remark this answers is not in the document]
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:442]: I could force an artificial rendered item to wait for it naturally. But it seemed even hackier than the wait 😂

File: server/app/playbook_run_service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply server/app/playbook_run_service.go:2721]: It's more of the behavior of being able to message a channel you are not a member of. Not sure if there is a precedent for that anywhere in the system. 

Technically you could change your username into the message you want sent. 

I don't think its too much of an issue since the offending user is easily identified.

[the remark this answers is not in the document]
Comment (trilopin) [reply server/app/playbook_run_service.go:2721]: now I get it, thanks for the explanation. I will raise this topic in order to see if we can avoid that scenario.

**Review by abhijit-singh — CHANGES_REQUESTED** [review]
Thanks, @trilopin! The flows look great! I just found 1 functional issue:
1. When I start a run, and then navigate to the run detail page, I still see the `Get involved` button. This should not be the case because I'm already a member and owner of that run. 



Also a couple of minor visual issues. These are non-blocking and can also be fixed with other smaller UX fixes...
1. Add 8px space between the get involved button and the info icon. 
2. For the `get involved` button, font-size should be 12 px.

**Comment by trilopin** [comment]
@agarciamontoro I fixed the flow problem that @abhijit-singh noticed in https://github.com/mattermost/mattermost-plugin-playbooks/pull/1295/commits/c796033f890c8d9f9fb259314e58b5ab2d8cb8f1 . However, please take a look because it seems a bit hacky to me.

**details**
the useChannel hook is built with useThing hook, which tries to get data from redux before doing a new fetch. If the data lacked some fields, we don't ask again for the full data.

The run we sent from server to webapp after Run creation was not the same as the Run we extracted from DB. 

My change: I forced the participantIds in run-creation to respect all Role-checks in the webapp side.  

I'm wondering if it's not better to force a complete run retrieval from DB, instead of hacking tons of things.

File: server/app/playbook_run_service.go [hunk 5]
```diff
@@ -474,6 +474,9 @@ func (s *PlaybookRunServiceImpl) CreatePlaybookRun(playbookRun *PlaybookRun, pb
 		}
 	}
 
+	// force owner as participant without triggering a store retrieval
+	playbookRun.ParticipantIDs = append(playbookRun.ParticipantIDs, playbookRun.OwnerUserID)
```
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply server/app/playbook_run_service.go:478]: Also, we should probably check if we do a similar append in the webapp side.

[the remark this answers is not in the document]
Comment (trilopin) [reply server/app/playbook_run_service.go:478]: ok, I'll be more defensive here

**Comment by trilopin** [comment]
> Thanks, @trilopin! The flows look great! I just found 1 functional issue:
> 
> 1. When I start a run, and then navigate to the run detail page, I still see the `Get involved` button. This should not be the case because I'm already a member and owner of that run.
>    
>          
>            [
>              
>                
>              
>            ](https://user-images.githubusercontent.com/14369607/178426375-bd9137a7-3b4f-45bc-a68c-5ce79dafc068.gif)
>            
>            
>              
>                
>                  
>                
>                
>                  
>                  
>                
>              
>              [
>                
>                  
>                
>              ](https://user-images.githubusercontent.com/14369607/178426375-bd9137a7-3b4f-45bc-a68c-5ce79dafc068.gif)

wow, very good catch

> 
> Also a couple of minor visual issues. These are non-blocking and can also be fixed with other smaller UX fixes...
> 
> 1. Add 8px space between the get involved button and the info icon.
> 2. For the `get involved` button, font-size should be 12 px.

Thanks a ton @abhijit-singh, I pushed some changes that should have fixed the style and flow issues.

File: server/api/playbook_runs.go [hunk 6]
```diff
@@ -516,7 +517,14 @@ func (h *PlaybookRunHandler) createPlaybookRun(playbookRun app.PlaybookRun, user
...
 	}
 
-	return h.playbookRunService.CreatePlaybookRun(&playbookRun, playbook, userID, public)
+	playbookRunReturned, err := h.playbookRunService.CreatePlaybookRun(&playbookRun, playbook, userID, public)
+	if err != nil {
+		return nil, err
+	}
+
+	// force database retrieval to ensure all data is processed correctly (i.e participantIds)
+	return h.playbookRunService.GetPlaybookRun(playbookRunReturned.ID)
```
Comment (agarciamontoro) [inline server/api/playbook_runs.go:526]: Much better, thanks!

**Review by agarciamontoro — APPROVED** [review]

**Review by abhijit-singh — APPROVED** [review]
Looks good!
